File: src/repositories/CaseRepository.ts
import { getDb } from "../db/database.ts";
import type {
  Case,
  CreateCaseInput,
  UpdateCaseInput,
  CaseStatus,
} from "../domains/cases/entities/Case.ts";
import {
  EncryptionService,
  type EncryptedData,
} from "../services/EncryptionService.ts";
import type { AuditLogger } from "../services/AuditLogger.ts";

export class CaseRepository {
  private encryptionService: EncryptionService;
  private auditLogger?: AuditLogger;

  constructor(encryptionService: EncryptionService, auditLogger?: AuditLogger) {
    this.encryptionService = encryptionService;
    this.auditLogger = auditLogger;
  }
  /**
   * Create a new case
   */
  create(input: CreateCaseInput): Case {
    try {
      const db = getDb();
      const encryption = this.requireEncryptionService();

      // Encrypt description before INSERT
      let descriptionToStore: string | null = null;
      if (input.description) {
        const encryptedDescription = encryption.encrypt(input.description);
        descriptionToStore = encryptedDescription
          ? JSON.stringify(encryptedDescription)
          : null;
      }

      const stmt = db.prepare(`
        INSERT INTO cases (title, description, case_type, status)
        VALUES (@title, @description, @caseType, 'active')
      `);

      const result = stmt.run({
        title: input.title,
        description: descriptionToStore,
        caseType: input.caseType,
      });

      const createdCase = this.findById(result.lastInsertRowid as number)!;

      // Audit: Case created
      this.auditLogger?.log({
        eventType: "case.create",
        resourceType: "case",
        resourceId: createdCase.id.toString(),
        action: "create",
        details: {
          title: createdCase.title,
          caseType: createdCase.caseType,
        },
        success: true,
      });

      return createdCase;
    } catch (error) {
      // Audit: Failed case creation
      this.auditLogger?.log({
        eventType: "case.create",
        resourceType: "case",
        resourceId: "unknown",
        action: "create",
        success: false,
        errorMessage: error instanceof Error ? error.message : "Unknown error",
      });
      throw error;
    }
  }

  /**
   * Find case by ID
   */
  findById(id: number): Case | null {
    const db = getDb();
    const stmt = db.prepare(`
      SELECT
        id,
        title,
        description,
        case_type as caseType,
        status,
        user_id as userId,
        created_at as createdAt,
        updated_at as updatedAt
      FROM cases
      WHERE id = ?
    `);

    const row = stmt.get(id) as Case | null;

    if (row) {
      // Decrypt description after SELECT
      const originalDescription = row.description;
      row.description = this.decryptDescription(row.description);

      // Audit: PII accessed (encrypted description field)
      if (originalDescription && row.description !== originalDescription) {
        this.auditLogger?.log({
          eventType: "case.pii_access",
          resourceType: "case",
          resourceId: id.toString(),
          action: "read",
          details: { field: "description", encrypted: true },
          success: true,
        });
      }
    }

    return row ?? null;
  }

  /**
   * Find all cases belonging to a specific user
   */
  findByUserId(userId: number): Case[] {
    const db = getDb();

    const query = `
      SELECT
        id,
        title,
        description,
        case_type as caseType,
        status,
        user_id as userId,
        created_at as createdAt,
        updated_at as updatedAt
      FROM cases
      WHERE user_id = ?
    `;

    const rows = db.prepare(query).all(userId) as Case[];

    // Decrypt descriptions if encryption service is available
    return rows.map((row) => {
      let description: string | null = row.description;

      if (description && this.encryptionService) {
        try {
          const encryptedData = JSON.parse(description) as EncryptedData;
          if (this.encryptionService.isEncrypted(encryptedData)) {
            description = this.encryptionService.decrypt(encryptedData);
          }
        } catch {
          // Legacy plaintext or decryption failure - keep as-is
        }
      }

      return {
        ...row,
        description,
      };
    });
  }

  /**
   * Find all cases with optional status filter
   */
  findAll(status?: CaseStatus): Case[] {
    const db = getDb();

    let query = `
      SELECT
        id,
        title,
        description,
        case_type as caseType,
        status,
        user_id as userId,
        created_at as createdAt,
        updated_at as updatedAt
      FROM cases
    `;

    let rows: Case[];

    if (status) {
      query += " WHERE status = ?";
      rows = db.prepare(query).all(status) as Case[];
    } else {
      rows = db.prepare(query).all() as Case[];
    }

    // Use batch decryption if enabled and encryption service is available
    const useBatchEncryption = process.env.ENABLE_BATCH_ENCRYPTION !== "false";

    if (useBatchEncryption && this.encryptionService && rows.length > 0) {
      const encryptionService = this.encryptionService;
      // Collect all encrypted descriptions for batch decryption
      const encryptedDescriptions = rows.map((row) => {
        if (!row.description) {
          return null;
        }

        try {
          const encryptedData = JSON.parse(row.description) as EncryptedData;
          return encryptionService.isEncrypted(encryptedData)
            ? encryptedData
            : null;
        } catch {
          return null; // Legacy plaintext
        }
      });

      try {
        // Batch decrypt all encrypted descriptions
        const decryptedDescriptions = encryptionService.batchDecrypt(
          encryptedDescriptions
        );

        // Map decrypted descriptions back to rows
        return rows.map((row, index) => {
          let description: string | null = row.description;

          // If we have a decrypted value from batch, use it
          if (encryptedDescriptions[index] !== null) {
            description = decryptedDescriptions[index];
          } else if (row.description && !encryptedDescriptions[index]) {
            // Legacy plaintext or failed parse - keep original
            description = row.description;
          }

          return {
            ...row,
            description,
          };
        });
      } catch (error) {
        // Graceful fallback for legacy or corrupted entries
        const errorMessage =
          error instanceof Error ? error.message : "Unknown error";
        this.auditLogger?.log({
          eventType: "encryption.decrypt",
          resourceType: "case",
          resourceId: "batch",
          action: "decrypt",
          details: {
            count: rows.length,
            reason: "batch_decrypt_failed",
            strategy: "batch_fallback",
          },
          success: false,
          errorMessage,
        });

        return rows.map((row) => ({
          ...row,
          description: this.decryptDescription(row.description),
        }));
      }
    }

    // Fallback to individual decryption
    return rows.map((row) => ({
      ...row,
      description: this.decryptDescription(row.description),
    }));
  }

  /**
   * Update case
   */
  update(id: number, input: UpdateCaseInput): Case | null {
    try {
      const db = getDb();
      const encryption = this.requireEncryptionService();

      const updates: string[] = [];
      const params: Record<string, unknown> = { id };

      if (input.title !== undefined) {
        updates.push("title = @title");
        params.title = input.title;
      }
      if (input.description !== undefined) {
        updates.push("description = @description");
        // Encrypt description before UPDATE
        if (input.description) {
          const encryptedDescription = encryption.encrypt(input.description);
          params.description = encryptedDescription
            ? JSON.stringify(encryptedDescription)
            : null;
        } else {
          params.description = null;
        }
      }
      if (input.caseType !== undefined) {
        updates.push("case_type = @caseType");
        params.caseType = input.caseType;
      }
      if (input.status !== undefined) {
        updates.push("status = @status");
        params.status = input.status;
      }

      if (updates.length === 0) {
        return this.findById(id);
      }

      const stmt = db.prepare(`
        UPDATE cases
        SET ${updates.join(", ")}
        WHERE id = @id
      `);

      stmt.run(params);

      const updatedCase = this.findById(id);

      // Audit: Case updated
      this.auditLogger?.log({
        eventType: "case.update",
        resourceType: "case",
        resourceId: id.toString(),
        action: "update",
        details: {
          fieldsUpdated: Object.keys(input),
        },
        success: true,
      });

      return updatedCase;
    } catch (error) {
      // Audit: Failed update
      this.auditLogger?.log({
        eventType: "case.update",
        resourceType: "case",
        resourceId: id.toString(),
        action: "update",
        success: false,
        errorMessage: error instanceof Error ? error.message : "Unknown error",
      });
      throw error;
    }
  }

  /**
   * Delete case (cascades to related records via foreign keys)
   */
  delete(id: number): boolean {
    try {
      const db = getDb();
      const stmt = db.prepare("DELETE FROM cases WHERE id = ?");
      const result = stmt.run(id);
      const success = result.changes > 0;

      // Audit: Case deleted
      this.auditLogger?.log({
        eventType: "case.delete",
        resourceType: "case",
        resourceId: id.toString(),
        action: "delete",
        success,
      });

      return success;
    } catch (error) {
      // Audit: Failed deletion
      this.auditLogger?.log({
        eventType: "case.delete",
        resourceType: "case",
        resourceId: id.toString(),
        action: "delete",
        success: false,
        errorMessage: error instanceof Error ? error.message : "Unknown error",
      });
      throw error;
    }
  }

  /**
   * Close a case
   */
  close(id: number): Case | null {
    return this.update(id, { status: "closed" });
  }

  /**
   * Get case count by status
   */
  countByStatus(): Record<CaseStatus, number> {
    const db = getDb();
    const stmt = db.prepare(`
      SELECT status, COUNT(*) as count
      FROM cases
      GROUP BY status
    `);

    const results = stmt.all() as Array<{ status: CaseStatus; count: number }>;

    const counts: Record<CaseStatus, number> = {
      active: 0,
      closed: 0,
      pending: 0,
    };

    results.forEach((row) => {
      counts[row.status] = row.count;
    });

    return counts;
  }

  /**
   * Get case statistics (total count + status breakdown)
   */
  getStatistics(): {
    totalCases: number;
    statusCounts: Record<CaseStatus, number>;
  } {
    const statusCounts = this.countByStatus();
    const totalCases =
      statusCounts.active + statusCounts.closed + statusCounts.pending;

    return {
      totalCases,
      statusCounts,
    };
  }

  /**
   * Decrypt description field with backward compatibility
   * @param storedValue - Encrypted JSON string or legacy plaintext
   * @returns Decrypted plaintext or null
   */
  private decryptDescription(
    storedValue: string | null | undefined
  ): string | null {
    if (!storedValue) {
      return null;
    }

    // If no encryption service, return as-is (backward compatibility)
    if (!this.encryptionService) {
      return storedValue;
    }

    try {
      // Try to parse as encrypted data
      const encryptedData = JSON.parse(storedValue) as EncryptedData;

      // Verify it's actually encrypted data format
      if (this.encryptionService.isEncrypted(encryptedData)) {
        return this.encryptionService.decrypt(encryptedData);
      }

      // If it's not encrypted format, treat as legacy plaintext
      return storedValue;
    } catch (_error) {
      // JSON parse failed - likely legacy plaintext data
      return storedValue;
    }
  }

  private requireEncryptionService(): EncryptionService {
    if (!this.encryptionService) {
      throw new Error("EncryptionService not configured for CaseRepository");
    }
    return this.encryptionService;
  }

  /**
   * Search cases by query string and filters
   */
  async searchCases(
    userId: number,
    query: string,
    filters?: any
  ): Promise<Case[]> {
    const db = getDb();
    const conditions: string[] = [];
    const params: any[] = [];

    // User filter
    conditions.push("user_id = ?");
    params.push(userId);

    // Text search
    if (query) {
      conditions.push("(title LIKE ? OR description LIKE ?)");
      params.push(`%${query}%`, `%${query}%`);
    }

    // Status filter
    if (filters?.caseStatus && filters.caseStatus.length > 0) {
      const placeholders = filters.caseStatus.map(() => "?").join(",");
      conditions.push(`status IN (${placeholders})`);
      params.push(...filters.caseStatus);
    }

    // Date range filter
    if (filters?.dateRange) {
      conditions.push("created_at >= ? AND created_at <= ?");
      params.push(
        filters.dateRange.from.toISOString(),
        filters.dateRange.to.toISOString()
      );
    }

    const whereClause =
      conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";

    const stmt = db.prepare(`
      SELECT
        id,
        title,
        description,
        case_type as caseType,
        status,
        user_id as userId,
        created_at as createdAt,
        updated_at as updatedAt
      FROM cases
      ${whereClause}
      ORDER BY created_at DESC
    `);

    const rows = stmt.all(...params) as Case[];

    // Decrypt descriptions
    return rows.map((row) => {
      row.description = this.decryptDescription(row.description);
      return row;
    });
  }

  /**
   * Get cases by user ID
   */
  async getByUserId(userId: number): Promise<Case[]> {
    const db = getDb();
    const stmt = db.prepare(`
      SELECT
        id,
        title,
        description,
        case_type as caseType,
        status,
        user_id as userId,
        created_at as createdAt,
        updated_at as updatedAt
      FROM cases
      WHERE user_id = ?
      ORDER BY created_at DESC
    `);

    const rows = stmt.all(userId) as Case[];

    // Decrypt descriptions
    return rows.map((row) => {
      row.description = this.decryptDescription(row.description);
      return row;
    });
  }

  /**
   * Get case by ID (async version for consistency)
   */
  async get(id: number): Promise<Case | null> {
    return this.findById(id);
  }
}

File: backend/repositories/case_repository.py
"""
Case repository for legal case management.
Ported from src/repositories/CaseRepository.ts (573 lines)

Features:
- Field-level encryption for description field (AES-256-GCM)
- Backward compatibility with legacy plaintext data
- Batch decryption for performance optimization
- Comprehensive audit logging
- Full CRUD operations
- Advanced search with filters
- Case statistics
"""

import json
from typing import List, Optional, Dict, Any

from sqlalchemy import func
from sqlalchemy.orm import Session, make_transient

from backend.models.case import Case, CaseStatus
from backend.schemas.case import CaseCreate, CaseUpdate
from backend.services.encryption_service import EncryptionService, EncryptedData
from backend.services.audit_logger import AuditLogger


class CaseRepository:
    def __init__(
        self,
        db: Session,
        encryption_service: EncryptionService,
        audit_logger: Optional[AuditLogger] = None,
    ):
        self.db = db
        self.encryption_service = encryption_service
        self.audit_logger = audit_logger

    def _encrypt_description(self, description: Optional[str]) -> Optional[str]:
        if not description:
            return None
        encrypted_data = self.encryption_service.encrypt(description)
        return json.dumps(encrypted_data.to_dict()) if encrypted_data else None

    def _decrypt_description(self, description: Optional[str]) -> Optional[str]:
        if not description:
            return None
        try:
            encrypted_dict = json.loads(description)
            # Check if it looks like our encrypted data structure
            if (
                isinstance(encrypted_dict, dict)
                and "ciphertext" in encrypted_dict
                and "iv" in encrypted_dict
            ):
                encrypted_data = EncryptedData.from_dict(encrypted_dict)
                return self.encryption_service.decrypt(encrypted_data)
            return description  # Not our format, return as is
        except (json.JSONDecodeError, TypeError, KeyError):
            # If it's not valid JSON or doesn't match EncryptedData format,
            # assume it's not encrypted or corrupted.
            return description

    def _detach_and_decrypt(self, case: Case) -> Case:
        """Detach from session and decrypt description to avoid type errors."""
        self.db.expunge(case)
        make_transient(case)
        description_value = getattr(case, "description", None)
        decrypted_description = self._decrypt_description(description_value)
        setattr(case, "description", decrypted_description)
        return case

    def create_case(self, case_data: CaseCreate, user_id: str) -> Case:
        """
        Create a new case.

        Args:
            case_data: CaseCreate with case details
            user_id: User ID

        Returns:
            Created Case model with decrypted description

        Raises:
            RuntimeError: If encryption or database operation fails
        """
        try:
            encrypted_description = self._encrypt_description(case_data.description)

            # Create case model
            db_case = Case(
                title=case_data.title,
                description=encrypted_description,
                case_type=case_data.case_type,
                status=CaseStatus.ACTIVE,
                user_id=user_id,
            )

            self.db.add(db_case)
            self.db.commit()
            self.db.refresh(db_case)

            # Audit: Case created
            if self.audit_logger:
                self.audit_logger.log(
                    event_type="case.create",
                    user_id=user_id,
                    resource_type="case",
                    resource_id=str(db_case.id),
                    action="create",
                    details={"title": db_case.title, "case_type": case_data.case_type},
                    success=True,
                )

            # Decrypt description for return value
            return self._detach_and_decrypt(db_case)

        except Exception as error:
            # Audit: Failed case creation
            if self.audit_logger:
                self.audit_logger.log(
                    event_type="case.create",
                    user_id=user_id,
                    resource_type="case",
                    resource_id="unknown",
                    action="create",
                    details={},
                    success=False,
                    error_message=str(error),
                )
            raise RuntimeError(f"Failed to create case: {str(error)}")

    def get_case_by_id(
        self, case_id: int, user_id: Optional[str] = None, decrypt: bool = False
    ) -> Optional[Case]:
        """
        Find case by ID.

        Args:
            case_id: Case ID
            user_id: Optional user ID for auditing
            decrypt: Whether to decrypt the description

        Returns:
            Case model with decrypted description or None if not found
        """
        case = self.db.query(Case).filter(Case.id == case_id).first()
        if case and decrypt:
            return self._detach_and_decrypt(case)
        return case

    def get_all_cases(self, skip: int = 0, limit: int = 100, decrypt: bool = False) -> List[Case]:
        """
        Get all cases with optional pagination and decryption.

        Args:
            skip: Number of records to skip (for pagination)
            limit: Maximum number of records to return
            decrypt: Whether to decrypt the description fields

        Returns:
            List of Case models
        """
        cases = self.db.query(Case).offset(skip).limit(limit).all()
        if decrypt:
            return [self._detach_and_decrypt(case) for case in cases]
        return cases

    def update_case(self, case_id: int, input_data: CaseUpdate, user_id: str) -> Optional[Case]:
        """
        Update case.

        Args:
            case_id: Case ID
            input_data: CaseUpdate with fields to update
            user_id: User ID

        Returns:
            Updated Case model with decrypted description or None if not found

        Raises:
            RuntimeError: If encryption or database operation fails
        """
        try:
            case = self.db.query(Case).filter(Case.id == case_id).first()

            if not case:
                return None

            update_data = input_data.model_dump(exclude_unset=True)
            updated_fields = []

            if "title" in update_data:
                case.title = update_data["title"]
                updated_fields.append("title")

            if "description" in update_data:
                setattr(case, "description", self._encrypt_description(update_data["description"]))
                updated_fields.append("description")

            if updated_fields:
                self.db.commit()
                self.db.refresh(case)

            if self.audit_logger and updated_fields:
                self.audit_logger.log(
                    event_type="case.update",
                    user_id=user_id,
                    resource_type="case",
                    resource_id=str(case_id),
                    action="update",
                    details={"updated_fields": updated_fields},
                    success=True,
                )

            # Decrypt for return
            return self._detach_and_decrypt(case)

        except Exception as error:
            # Audit: Failed update
            if self.audit_logger:
                self.audit_logger.log(
                    event_type="case.update",
                    user_id=user_id,
                    resource_type="case",
                    resource_id=str(case_id),
                    action="update",
                    details={},
                    success=False,
                    error_message=str(error),
                )
            raise RuntimeError(f"Failed to update case: {str(error)}")

    def delete_case(self, case_id: int, user_id: str) -> bool:
        """
        Delete case (cascades to related records via foreign keys).

        Args:
            case_id: Case ID
            user_id: User ID for auditing

        Returns:
            True if case was deleted, False if not found
        """
        try:
            case = self.db.query(Case).filter(Case.id == case_id).first()

            if not case:
                success = False
            else:
                self.db.delete(case)
                self.db.commit()
                success = True

            # Audit: Case deleted
            if self.audit_logger:
                self.audit_logger.log(
                    event_type="case.delete",
                    user_id=user_id,
                    resource_type="case",
                    resource_id=str(case_id),
                    action="delete",
                    details={},
                    success=success,
                )

            return success

        except Exception as error:
            # Audit: Failed deletion
            if self.audit_logger:
                self.audit_logger.log(
                    event_type="case.delete",
                    user_id=user_id,
                    resource_type="case",
                    resource_id=str(case_id),
                    action="delete",
                    details={},
                    success=False,
                    error_message=str(error),
                )
            raise RuntimeError(f"Failed to delete case: {str(error)}")

    def count_by_status(self) -> Dict[str, int]:
        """
        Get case count by status.

        Returns:
            Dictionary mapping status to count
        """
        results = (
            self.db.query(Case.status, func.count(Case.id).label("count"))
            .group_by(Case.status)
            .all()
        )

        counts = {"active": 0, "closed": 0, "pending": 0}

        for status, count in results:
            if status in counts:
                counts[status] = count

        return counts

    def search_cases(
        self, user_id: int, query: str, filters: Optional[Dict[str, Any]] = None
    ) -> List[Case]:
        """
        Search cases by query string and filters.

        Args:
            user_id: User ID
            query: Search query string
            filters: Optional filters dictionary with:
                - case_status: List of status strings
                - date_range: Dict with "from" and "to" datetime objects

        Returns:
            List of Case models with decrypted descriptions
        """
        # Start with user filter
        query_builder = self.db.query(Case).filter(Case.user_id == user_id)

        # Text search (NOTE: This searches encrypted descriptions, so won't match)
        # For proper search, you'd need to decrypt all descriptions first
        if query:
            query_builder = query_builder.filter((Case.title.like(f"%{query}%")))

        # Status filter
        if filters and filters.get("case_status"):
            case_statuses = [CaseStatus(s) for s in filters["case_status"]]
            query_builder = query_builder.filter(Case.status.in_(case_statuses))

        # Date range filter
        if filters and filters.get("date_range"):
            date_range = filters["date_range"]
            query_builder = query_builder.filter(
                Case.created_at >= date_range["from"], Case.created_at <= date_range["to"]
            )

        # Execute query
        cases = query_builder.order_by(Case.created_at.desc()).all()

        # Decrypt descriptions
        return [self._detach_and_decrypt(case) for case in cases]

    def get_by_user_id(self, user_id: int) -> List[Case]:
        """
        Get cases by user ID (async version for consistency).

        Args:
            user_id: User ID

        Returns:
            List of Case models with decrypted descriptions
        """
        cases = (
            self.db.query(Case)
            .filter(Case.user_id == user_id)
            .order_by(Case.created_at.desc())
            .all()
        )

        # Decrypt descriptions
        return [self._detach_and_decrypt(case) for case in cases]

    def get(self, case_id: int) -> Optional[Case]:
        """
        Get case by ID (async version for consistency).

        Args:
            case_id: Case ID

        Returns:
            Case model with decrypted description or None if not found
        """
        return self.get_case_by_id(case_id, decrypt=True)

    # Private helper methods

    def _decrypt_case(self, case: Case) -> Case:
        """
        Decrypt a single case's description.

        Args:
            case: Case model with potentially encrypted description

        Returns:
            Case model with decrypted description
        """
        return self._detach_and_decrypt(case)

    def _require_encryption_service(self) -> EncryptionService:
        """
        Require encryption service to be configured.

        Returns:
            EncryptionService instance

        Raises:
            RuntimeError: If encryption service is not configured
        """
        if not self.encryption_service:
            raise RuntimeError("EncryptionService not configured for CaseRepository")
        return self.encryption_service
